modules/call-dispatch/src/lib.rs

```diff
@@ -60,75 +105,117 @@ pub trait Trait: frame_system::Trait {
 		/// Message has been dispatched with given result.
 		MessageDispatched(InstanceId, MessageId, DispatchResult),
+		/// Phantom member, never used.
+		Dummy(PhantomData<I>),
 	}
 );
 
 decl_module! {
 	/// Call Dispatch FRAME Pallet.
-	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+	pub struct Module<T: Trait<I>, I: Instance = DefaultInstance> for enum Call where origin: T::Origin {
 		/// Deposit one of this module's events by using the default implementation.
 		fn deposit_event() = default;
 	}
 }
 
-impl<T: Trait> MessageDispatch<T::MessageId> for Module<T> {
-	type Message = (SpecVersion, Weight, <T as Trait>::Call);
+impl<T: Trait<I>, I: Instance> MessageDispatch<T::MessageId> for Module<T, I> {
+	type Message = MessagePayload<
+		T::SourceChainAccountPublic,
+		T::TargetChainAccountPublic,
+		T::TargetChainSignature,
+		<T as Trait<I>>::Call,
+	>;
 
 	fn dispatch(bridge: InstanceId, id: T::MessageId, message: Self::Message) -> Weight {
-		let (spec_version, weight, call) = message;
-
 		// verify spec version
 		// (we want it to be the same, because otherwise we may decode Call improperly)
 		let expected_version = <T as frame_system::Trait>::Version::get().spec_version;
-		if spec_version != expected_version {
+		if message.spec_version != expected_version {
 			frame_support::debug::trace!(
 				"Message {:?}/{:?}: spec_version mismatch. Expected {:?}, got {:?}",
 				bridge,
 				id,
 				expected_version,
-				spec_version,
+				message.spec_version,
 			);
-			Self::deposit_event(Event::<T>::MessageVersionSpecMismatch(
+			Self::deposit_event(RawEvent::MessageVersionSpecMismatch(
 				bridge,
 				id,
 				expected_version,
-				spec_version,
+				message.spec_version,
 			));
 			return DEPOSIT_EVENT_WEIGHT;
 		}
 
 		// verify weight
 		// (we want passed weight to be at least equal to pre-dispatch weight of the call
 		// because otherwise Calls may be dispatched at lower price)
-		let dispatch_info = call.get_dispatch_info();
+		let dispatch_info = message.call.get_dispatch_info();
 		let expected_weight = dispatch_info.weight;
-		if weight < expected_weight {
+		if message.weight < expected_weight {
 			frame_support::debug::trace!(
 				"Message {:?}/{:?}: passed weight is too low. Expected at least {:?}, got {:?}",
 				bridge,
 				id,
 				expected_weight,
-				weight,
+				message.weight,
 			);
-			Self::deposit_event(Event::<T>::MessageWeightMismatch(bridge, id, expected_weight, weight));
+			Self::deposit_event(RawEvent::MessageWeightMismatch(
+				bridge,
+				id,
+				expected_weight,
+				message.weight,
+			));
 			return DEPOSIT_EVENT_WEIGHT;
 		}
 
+		// prepare dispatch origin
+		let origin_account = match message.origin {
+			CallOrigin::BridgeAccount => bridge_account_id(bridge, CALL_DISPATCH_MODULE_PREFIX),
+			CallOrigin::RealAccount(source_public, target_public, target_signature) => {
+				let mut signed_message = Vec::new();
+				message.call.encode_to(&mut signed_message);
```

Comment:
I was thinking that `spec_version` might be required to be here as well, but after a deeper thought it's not needed, because:
1. We verify the whole `MessagePayload` objects indirectly, by verifying the proof of inclusion on the target chain (i.e. no tampering possible here)
2. Since we add `source_public` to the `signed_message` here, it's only possible for the `source_public` account on the source chain to add messages to the outbound queue.
3. So it means that only the original sender can replay the same `origin(with signature) + call` on the source chain against a different `spec_version` or `weight`.